core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,255 @@
+  }
+
+  public class Tracker extends ServiceTracker<Object, Object> {
+
+    private final LinkedHashMap<Class<?>, Object> registeredServices;
+
+    public Tracker(
+        BundleContext context, Filter filter, LinkedHashMap<Class<?>, Object> registeredServices) {
+      super(context, filter, null);
+      this.registeredServices = registeredServices;
+    }
+
+    /**
+     * It will try to match all dependentServices with a ServiceReference notification. After the
+     * notification is handled, it checks if all dependentServices are not null. If they are not,
+     * the client's provided {@link this#createService(List)} is called, and the service is
+     * registered. It will not register the service if it was already registered.
+     */
+    @Override
+    public Object addingService(ServiceReference<Object> ref) {
+      Object service = super.addingService(ref);
+      startIfAllRegistered(ref, service);
+
+      return service;
+    }
+
+    @VisibleForTesting
+    public void startIfAllRegistered(ServiceReference<Object> ref, Object service) {
+      if (service == null) {
+        return;
+      }
+      for (Map.Entry<Class<?>, Object> registeredService : registeredServices.entrySet()) {
+        if (registeredService.getKey().isAssignableFrom(service.getClass())) {
+          logger.debug("{} using service: {}", activatorName, ref.getBundle());
+          registeredServices.put(registeredService.getKey(), service);
+        }
+      }
+      if (registeredServices.values().stream().allMatch(Objects::nonNull)) {
+        startServiceInternal(new LinkedList<>(registeredServices.values()));
+      }
+    }
+  }
+
+  /**
+   * Clients should override this method to create the Service that will be registered in the OSGi
+   * container. The dependant services will contain all services registered passed to the
+   * constructor of this class as {@code List<Class<?>> dependentServices}. The ordering will be the
+   * same. You can safely cast the objects to the expected types according to the dependentServices.
+   *
+   * @return ServiceAndProperties that has the service for OSGi registration and the properties that
+   *     will be passed or null if there is no registration service required.
+   */
+  @Nullable
+  protected abstract ServiceAndProperties createService(List<Object> dependentServices);
+
+  /**
+   * It will be called when the OSGi calls {@link this#stop(BundleContext)} and only if service was
+   * already started.
+   */
+  protected abstract void stopService();
+
+  public static class ServiceAndProperties {
+    private final Object service;
+
+    private final Hashtable<String, String> properties;
+
+    public ServiceAndProperties(Object service, Hashtable<String, String> properties) {
+      this.service = service;
+      this.properties = properties;
+    }
+  }
+
+  public static class DependentService {
+    private Class<?> className;
+
+    private Optional<String> identifier;
```

Comment:
How about allowing any OSGi filter for generalizability?

Reply:
I am not sure what you mean. The filter is used at a different level, see:
`new Tracker(
            context, context.createFilter(constructDependenciesFilter()), registeredServices);`
and it is constructed from a `String` (we are doing this for all other activators).

Reply:
I mean allowing BaseActivator sub-classes to provide a general-purpose OSGI filter, not just the service identifier.

Reply:
I see your point but I don’t see a use case for it. We have only two cases: exact class and identifier. Both are supported by the current solution. If there will be a new use case, we can easily parameterize it or expose filter.

Reply:
ok